src/Utils/LexicalDistance.php

```diff
@@ -0,0 +1,133 @@
+<?php declare(strict_types=1);
+
+namespace GraphQL\Utils;
+
+/**
+ * Computes the lexical distance between strings A and B.
+ *
+ * The "distance" between two strings is given by counting the minimum number
+ * of edits needed to transform string A into string B. An edit can be an
+ * insertion, deletion, or substitution of a single character, or a swap of two
+ * adjacent characters.
+ *
+ * Includes a custom alteration from Damerau-Levenshtein to treat case changes
+ * as a single edit which helps identify mis-cased values with an edit distance
+ * of 1.
+ *
+ * This distance can be useful for detecting typos in input or sorting
+ */
+class LexicalDistance
+{
+    private string $input;
+
+    private string $inputLowerCase;
+
+    /**
+     * @var array<int>
+     */
+    private array $inputArray;
+
+    /**
+     * @var array<array<int>>
+     */
+    private array $rows;
+
+    public function __construct(string $input)
+    {
+        $this->input = $input;
+        $this->inputLowerCase = \strtolower($input);
+        $this->inputArray = self::stringToArray($this->inputLowerCase);
+
+        $length = \mb_strlen($input);
+        $this->rows = [
+            \array_fill(0, $length, 0),
+            \array_fill(0, $length, 0),
+            \array_fill(0, $length, 0),
+        ];
```

Comment:
Do we need to do this, given PHP can auto-populate intermediary arrays when populating nested values?

Reply:
My first impression was that it would lead to `??` in the code but no, you are right. It's enough to initialise the `rows` property with an empty array.

I'm now wondering if sharing the `rows` array between multiple `measure()` calls actually makes sense at all and if a local `$rows` variable initialised in the place of `$rows = &$this->rows;` wouldn't be better. But I'm leaving the property there for now in order not to divert too much for the JS code.

Reply:
I would rater see it simplified. The JS code seems to be optimized for performance in a way that does not apply to PHP - in such cases I opted to go for optimized PHP code instead. The biggest value the reference implementation provides is the test suite and a starting point for the implementation.